### PR TITLE
Add cloudpathlib

### PR DESCRIPTION
**Pull request by jayqi**
Checklist
- [X] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [X] Source is from official source
- [X] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [X] If static libraries are linked in, the license of the static library is packaged.
- [X] Build number is 0
- [X] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [X] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

Confirming that I am willing to be listed as a maintainer.

### CODE REVIEW

**Comment by jayqi** [comment]
Hi conda-forge maintainers. I've been struggling a bit trying to configure optional dependencies, following the example of [matplotlib's feedstock](https://github.com/conda-forge/matplotlib-feedstock/blob/master/recipe/meta.yaml).

For PyPI, this package has "extras" defined for `cloudpathlib[s3]` and `cloudpathlib[azure]`. I'm trying to do something similar with subpackages. Any pointers are appreciated. 

I'm currently getting failures that look like 

```
INFO conda_build.metadata:finalize_outputs_pass(707): Attempting to finalize metadata for cloudpathlib-base
Error: Failed to render jinja template in /home/conda/staged-recipes-copy/recipes/cloudpathlib/meta.yaml:
'PYTHON' is undefined
```

File: recipes/cloudpathlib/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+  - name: cloudpathlib-azure
+    requirements:
+      host:
+        - python
+        - pip
+      run:
+        - python
+        - azure-storage-blob >=12
+        - {{ pin_subpackage('cloudpathlib-base', exact=True) }}
```
Comment (scopatz) [inline recipes/cloudpathlib/meta.yaml:45]: This should probably be a regular dep

File: recipes/cloudpathlib/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+test:
+  imports:
+    - cloudpathlib
+  commands:
+    - pip check
+  requires:
+    - pip
+
+outputs:
+  - name: cloudpathlib
```
Comment (scopatz) [inline recipes/cloudpathlib/meta.yaml:35]: This output probably needs more....

**Comment by jayqi** [comment]
Hi @scopatz, I think I got everything working now, and it's ready for review.

Based on following the example of [matplotlib's feedstock](https://github.com/conda-forge/matplotlib-feedstock/blob/master/recipe/meta.yaml) and my reading of the [conda build documentation for subpackages](https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html?highlight=pin_subpackage#referencing-subpackages), I think `pin_subpackage` seems like the correct way to embed the dependencies. 

Thanks!

**Comment by jayqi** [comment]
Hi @conda-forge/help-python, this PR is ready for review. Thanks!

**Comment by jayqi** [comment]
Hi @scopatz @conda-forge/help-python, this PR is ready for review. Thanks!

**Comment by jayqi** [comment]
Hi @conda-forge/core, this PR is ready for review, and I haven't heard anything from my previous reviewer or the Python team in several days. Thanks.

**Pull request merged**